### ISSUE TITLE
access to the full project XML configuration/PropertyGroups?

### ISSUE THREAD

**Issue by benblo** [Fody/Fody#170]
I have some custom properties defined in PropertyGroup in my projects: is there a way to access those?
If possible, I'd like to avoid replicating them in FodyWeavers.xml (as seen here https://github.com/Fody/Fody/wiki/ModuleWeaver)

**Comment by SimonCropp** [comment]
Not atm. But you have access to the project path so you could poke into the project xml

**Comment by benblo** [comment]
Yes I suppose, but properties have to actually be evaluated, which isn't
trivial. I was wondering if accessing them as a key-value store was part of
msbuild. Fody is like, an agent of that pipeline right? I don't know the
terminology.

**Comment by SimonCropp** [comment]
can you outline what feature you are trying to build that needs this?

**Comment by Rohansi** [comment]
Being able to use values such as `$(Configuration)` would help a lot when trying to access output directories.

**Comment by SimonCropp** [comment]
@benblo @Rohansi you are both giving me hypotheticals. What are you actually trying to build that requires this?

**Comment by Rohansi** [comment]
I'm working on an addin that requires access to other assemblies. If any of them are part of the same solution it would be very nice if the user could use `$(Configuration)` to use the correct one instead of hardcoding it for one configuration.

I'm not sure what MSBuild exposes but a function to get a property value by name and a function to expand properties in a string should be enough for this. I found this after a bit of googling but I don't know how useful it would be: http://stackoverflow.com/questions/2772426/how-to-access-the-msbuild-s-properties-list-when-coding-a-custom-task

If that doesn't work maybe there's can be a way to manually pass properties to Fody in the project file?

**Comment by SimonCropp** [comment]
> I'm working on an addin that requires access to other assemblies.

You have access to both `@(ReferencePath)` and `@(ReferenceCopyLocalPaths)`
https://github.com/Fody/Fody/wiki/ModuleWeaver

Also what r u using the other assemblies for?

> If any of them are part of the same solution it would be very nice if the user could use $(Configuration) to use the correct one instead of hardcoding it for one configuration.

I dont understand that statement

> $(Configuration)

You have access to the compile time constants `$(DefineConstants)` https://github.com/Fody/Fody/wiki/ModuleWeaver wouldn't that suffice?

Just to be clear i am not against adding features. but the list of things what could be passed into a weaver is huge and the associated ongoing support and testing effort is several times larger than that. so i just want to ensure we are actually delivering real value and usable features befor going down that path. 

Is this thing you are trying to build open source?

**Comment by Rohansi** [comment]
> You have access to both `@(ReferencePath)` and `@(ReferenceCopyLocalPaths)`

Unfortunately this isn't what I need. I'm making something that replaces uses of some types with another one defined in a different assembly. While I could force users to have an existing reference to that assembly, I don't think it would be a nice solution.

>> If any of them are part of the same solution it would be very nice if the user could use $(Configuration) to use the correct one instead of hardcoding it for one configuration.

> I dont understand that statement

By default Visual Studio sets the output path for binaries to `bin/$(Configuration)/`. Something like this is used by most projects so you need access to it so you can select the proper binary. The behavior could match MSBuild very closely which would be very nice.

> You have access to the compile time constants `$(DefineConstants)`

This could work if you assume there is just a Debug and Release configuration but it's a pretty bad assumption to make.

> Is this thing you are trying to build open source?

Not yet, but I'm planning on putting it up soon.

**Comment by distantcam** [comment]
@Rohansi So is your problem solved by just passing in `$(Configuration)`? Do you need all the MSBuild properties?

**Comment by Rohansi** [comment]
I only really need `$(Configuration)` right now but if grabbing most/all of them isn't too difficult, why not?

**Comment by distantcam** [comment]
@Rohansi Because it may be too difficult. There might also be performance implications as I imagine evaluating all those properties may not be cheap.

**Comment by benblo** [comment]
Hey Simon, sorry for the lack of answer.
So our sln has 15 projects, that all include a common .target file, where
we include a bunch of stuff, mostly define constants and references to
common DLLs. All csprojs define a couple "master" properties before
including that .targets, which then pushes different properties accordingly.
So basically we have 2 categories of projects (editor-only and runtime,
this is for Unity, a game engine), but they all use the same weaver
(in-solution) and I need to know which category I'm dealing with in the
weaver.

You're right in that parsing the defines helps, I can detect some stuff,
though it's not as direct. What I couldn't I added to FodyWeavers.xml,
which basically replicates the master properties I talked about. At the
time I had to add this 15 times, it felt silly to duplicate an info that
was already in the csproj so I thought "am I missing something?", but it's
definitely not that bad, just a nice to have.

**Comment by SimonCropp** [comment]
@Rohansi the link you posted. that is actually misleading. it does not read the current properties from the context of msbuild. is reads the properties defined inside project on the file system. so other propertied passed into the msbuild engine can not be resolved that way.

**Comment by SimonCropp** [comment]
i will add $(Configuration) in the next version https://github.com/Fody/Fody/issues/179

**Comment by SimonCropp** [comment]
@benblo if you want to have a common set of settings then the weaver has access to the project path. you can walk up the tree and find a settings file at the solution level. or alternatively you can use the link you posted to extract the project file based properties out yourself.

**Comment by benblo** [comment]
I'm not the one who posted the link, so I didn't even an MSBuild API existed ;), but yeah that would work, thanks!
Having a shared settings file wouldn't really help though, as I still wouldn't know what project is of what type.

On a (vaguely) related topic, is there any way to pass information between weavers of different projects (besides leaving files around I mean)? I'm not very familiar with msbuild tasks but I assume weavers are created/destroyed for each project in sequence. So I'm thinking something like having a solution-wide weaver, or maybe weavers having access to some shared data or list of weavers that ran before them...
The usecase is this: I use Fody (among other things) to generate files for every types in the project. So if I move a class from one project to another, I also need to move the generated file. At the moment one weaver destroys it (doesn't belong to it anymore), the next recreates it at another location, but this causes problems (versioning, Unity's file id changes).
Any ideas? Ideally I would collect a list of stuff to do, then when all projects are done building, run a final task to sort it all out and apply the changes in a more clever way.

**Comment by ilengyel** [comment]
I am looking to have access to the $(OutputPath) property - not always set to the default 'bin\$(configuratoin)'. Essentially we want to push additional files to the same directory of the weaving target. It looks tempting to pass the full list of evaluated properties..

**Comment by dekenless** [comment]
...and the $(OutDir) would be very helpful to have, as I am weaving based on a number of solutions.  Works fine locally when the outputs are under the project in which I am weaving, but when done on a build server this is not the case.
